File: SDK/Python SDK/servo_sdk/context.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal


Role = Literal["system", "user", "assistant"]


@dataclass
class Message:
    role: Role
    content: str


@dataclass
class Conversation:
    """
    Simple conversation buffer.

    The Servo backend currently accepts a single `prompt` string, so this object
    compiles message history into a single prompt.
    """

    system_prompt: str | None = None
    max_turns: int = 10
    messages: list[Message] = field(default_factory=list)

    def add_user(self, content: str) -> None:
        self.messages.append(Message(role="user", content=content))
        self._trim()

    def add_assistant(self, content: str) -> None:
        self.messages.append(Message(role="assistant", content=content))
        self._trim()

    def build_prompt(self, next_user_message: str | None = None) -> str:
        parts: list[str] = []
        if self.system_prompt:
            parts.append(f"System:\n{self.system_prompt}\n")

        for m in self.messages:
            label = "User" if m.role == "user" else "Assistant"
            parts.append(f"{label}:\n{m.content}\n")

        if next_user_message is not None:
            parts.append(f"User:\n{next_user_message}\n")

        return "\n".join(parts).strip()

    def _trim(self) -> None:
        if self.max_turns <= 0:
            return
        max_messages = self.max_turns * 2
        if len(self.messages) > max_messages:
            self.messages = self.messages[-max_messages:]

File: SDK/NPM SDK/src/context.ts
/**
 * Role types for messages in a conversation.
 */
export type Role = 'system' | 'user' | 'assistant';

/**
 * A single message in a conversation.
 */
export interface Message {
  role: Role;
  content: string;
}

/**
 * Simple conversation buffer for managing chat history.
 *
 * The Servo backend currently accepts a single `prompt` string,
 * so this class compiles message history into a formatted prompt.
 */
export class Conversation {
  private systemPrompt?: string;
  private maxTurns: number;
  private messages: Message[];

  constructor(systemPrompt?: string, maxTurns: number = 10) {
    this.systemPrompt = systemPrompt;
    this.maxTurns = maxTurns;
    this.messages = [];
  }

  /**
   * Add a user message to the conversation.
   */
  addUser(content: string): void {
    this.messages.push({ role: 'user', content });
    this.trim();
  }

  /**
   * Add an assistant message to the conversation.
   */
  addAssistant(content: string): void {
    this.messages.push({ role: 'assistant', content });
    this.trim();
  }

  /**
   * Build a formatted prompt string from the conversation history.
   */
  buildPrompt(nextUserMessage?: string): string {
    const parts: string[] = [];

    if (this.systemPrompt) {
      parts.push(`System:\n${this.systemPrompt}\n`);
    }

    for (const message of this.messages) {
      const label = message.role === 'user' ? 'User' : 'Assistant';
      parts.push(`${label}:\n${message.content}\n`);
    }

    if (nextUserMessage) {
      parts.push(`User:\n${nextUserMessage}\n`);
    }

    return parts.join('\n').trim();
  }

  /**
   * Trim messages to respect maxTurns limit.
   */
  private trim(): void {
    if (this.messages.length > this.maxTurns * 2) {
      this.messages = this.messages.slice(-this.maxTurns * 2);
    }
  }

  /**
   * Get all messages in the conversation.
   */
  getMessages(): Message[] {
    return [...this.messages];
  }

  /**
   * Clear all messages from the conversation.
   */
  clear(): void {
    this.messages = [];
  }
}
